### PR TITLE
Recreation

### PR DESCRIPTION
**Pull request by jawalonoski**
This pull request attempts to address #752.

Recreation of data requires the use of a random seed (`-s`), a clinician seed (`-cs`), and a reference date (`-r`).

Things that will not be identical:

- UUIDs: these will change across runs. Sorry.
- Export History: if you run the same parameters next year, and only keep the default 10 years of history, then the data in the record will be filtered differently (i.e. the oldest one year from the first run will not be exported)..

## Summary of Changes

- Randomness
  - I made the `Random` inside `Person` to be `private` and made a few new accessor methods (e.g. `randBoolean()`).
  - Except for a few places (e.g. `Generator.java`), that means `Random` objects are no longer passed around as method arguments, and there should be no use of `Random` in the code.
  - All random number access goes through the `Person#rand*` methods.
  - There should also be no `System.currentTimeInMillis()` calls except for a few spots (e.g. `Generator.java` and sometimes export paths when folder per run settings are used).
- Reference Dates
  - I added an optional `-r` switch for `referenceDate` which can be used to specify when Synthea bases all the birth dates from. In my testing I used `20200320` (or March 20th, 2020) but you can use any date. This defaults to now (`System.currentTimeinMillis()`) if no value is supplied.
- Modules and States
  - All modules and states are completely cloned across the patients now. The modules used to be passed directly, and only the states were cloned. Now the modules are cloned as well. This was done to remove any possibility of multi-thread race conditions on accessing modules. Possibly unnecessary.
- Wiki
  - I added this https://github.com/synthetichealth/synthea/wiki/Contributing#new-sources-of-randomness to the wiki.

### CODE REVIEW

File: src/main/java/org/mitre/synthea/engine/Module.java [hunk 1]
```diff
@@ -423,7 +440,7 @@ public synchronized Module get() {
       if (fault != null) {
         throw new RuntimeException(fault);
       }
-      return module;
+      return module.clone();
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/engine/Module.java:443]: Have you looked at how much this affects memory usage? The whole set of modules as raw text is 2 MB so broken out into full objects across multiple patients that might be a lot. If the real issue was the race condition in PediatricGrowthTrajectory and this increases memory usage by a lot, we may want to revert this piece.

Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/engine/Module.java:443]: image

This is a run of 1000 patients. My max memory is set to 8 GB.

Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/engine/Module.java:443]: image

Same run of 1000 patients from the latest `master`

File: src/main/java/org/mitre/synthea/world/agents/Person.java [hunk 2]
```diff
@@ -168,7 +168,7 @@ public Person(long seed) {
   }
 
   /**
-   * Retuns a random double.
+   * Returns a random double.
```
Comment (hadleynet) [inline src/main/java/org/mitre/synthea/world/agents/Person.java:171]: Rather than having the random number related methods as public Person methods, I think it would be cleaner if they were all gathered together into an interface (e.g. RandomProvider) that Person implements. Methods that need access to randomness but not other aspects of Person could then just accept a RandomProvider instead of Person and this would make it clear which methods need Person access vs just Randomness.

File: src/main/java/org/mitre/synthea/world/agents/Provider.java [hunk 3]
```diff
@@ -439,7 +439,8 @@ private Clinician generateClinician(long clinicianSeed, Random clinicianRand,
       long clinicianIdentifier, Provider provider) {
     Clinician clinician = null;
     try {
-      Demographics city = location.randomCity(clinicianRand);
+      Person doc = new Person(clinicianIdentifier);
```
Comment (hadleynet) [inline src/main/java/org/mitre/synthea/world/agents/Provider.java:442]: Should this be Person(clinicianSeed) instead?

File: src/main/java/org/mitre/synthea/engine/Module.java (hunk 1 above)
Comment (dehall) [reply src/main/java/org/mitre/synthea/engine/Module.java:443]: Thanks, doesn't seem like there's a meaningful difference

File: src/main/java/org/mitre/synthea/world/agents/Person.java (hunk 2 above)
Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/world/agents/Person.java:171]: An interface like RandomProvider is a good idea. I'll work on that.

File: src/main/java/org/mitre/synthea/world/agents/Provider.java (hunk 3 above)
Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/world/agents/Provider.java:442]: No. I admit that it is a little hard to follow, but `clinicianSeed` is the seed for ALL of the clinicians that can be passed in with the GeneratorOptions. The clinicianIdentifier is the seed for THIS particular clinician.

I'll make some changes in this area when I refactor with a `RandomProvider` as suggested above.

File: src/main/java/org/mitre/synthea/engine/Module.java [hunk 4]
```diff
@@ -284,6 +284,23 @@ public Module(JsonObject definition, boolean submodule) throws Exception {
...
 
+  /**
+   * Clone this module. Never provide the original.
+   */
+  public Module clone() {
+    Module clone = new Module();
+    clone.name = this.name;
+    clone.submodule = this.submodule;
+    clone.remarks = this.remarks;
+    clone.states = new ConcurrentHashMap<String, State>();
```
[the remark this answers is not in the document]
Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/engine/Module.java:295]: Good catch.

File: src/main/java/org/mitre/synthea/world/agents/Person.java (hunk 2 above)
Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/world/agents/Person.java:171]: I called the interface `RandomNumberGenerator` because I feel like `Provider` is overloaded with meaning, e.g. healthcare provider.

**Comment by jawalonoski** [comment]
@hadleynet Ready for your re-review.

**Comment by hadleynet** [comment]
Rebased and merged.
